lenskit-eval/src/main/java/org/lenskit/eval/traintest/DataSet.java

```diff
@@ -286,9 +306,31 @@ public static DataSetBuilder copyBuilder(DataSet data) {
      * @return The data source.
      */
     private static DataSet loadDataSet(JsonNode json, URI base, String name, int part) throws IOException {
+        System.out.println(json);
         Preconditions.checkArgument(json.has("train"), "%s: no train data specified", name);
         Preconditions.checkArgument(json.has("test"), "%s: no test data specified", name);
+
+        List<EntityType> entityList = new ArrayList();
+
+        if (json.has("entity_types")) {
```

Comment:
This would be easier to follow and less error-prone if we just look up `entity_types` once.

```java
JsonNode etNode = json.path("entity_types");
if (etNode.isArray()) {
    // it's an array, process accordingly (and no need to create an `arrayNode` variable, just use `etNode`)
} else if (etNode.isTextual()) {
    // it's textual
} else if (etNode.isMissing() || etNode.isNull()) {
    // missing or null, use the default
} else {
    throw new IllegalArgumentException("unexpected format for entity_types");
}
```